# TP NoSQL

Prénom : Laurent

Nom : Chen

Pour ce TP noté vous avez trois heures pour le réaliser et 48 heures pour le finir si besoin. Le rendu final est attendu pour dimanche 29 septembre 2024 à 23 h 59.

Vous serrez évalués sur les requêtes mais également sur la qualité et la propreté du code. Pour chaque question, une réponse présentant les résultats est attendue.

In [2]:
# installation des packages pandas et pymongo
!pip install -r requirements.txt --quiet

In [290]:
import pymongo
import pandas as pd

URI = 'mongodb+srv://user_mongo:a3ROig9CHznixVwL@cluster-but-sd.im7ug.mongodb.net/?retryWrites=true&w=majority&appName=cluster-but-sd'
client = pymongo.MongoClient(URI)
db = client.tp

# output the name of the collections in the database
print("Collections: ", db.list_collection_names())

Collections:  ['restaurants', 'movies']


1. Combien de films sont dans la base de données ?

In [10]:
db.movies.count_documents({})

23530

Il y a 23530 films dans la base de données.

2. Quel est le nombre moyen de votes par pays ? Trier les pays par nombre moyen de votes décroissant.

In [298]:
Q2 = db.movies.aggregate([
    { "$unwind": "$countries" },
    { "$group": { "_id": "$countries", "vote_moyen": { "$avg": "$imdb.votes" }}},
    { "$sort": { "vote_moyen": -1 }}
])
pd.DataFrame(list(Q2))

_id     vote_moyen
0                   Malta  210571.800000
1                 Bahamas  206061.000000
2             New Zealand   66589.269231
3    United Arab Emirates   53495.833333
4                     USA   35016.703663
..                    ...            ...
153                Kosovo     104.500000
154               Moldova     104.000000
155           North Korea      92.666667
156                Brunei      51.000000
157               Vanuatu       6.000000

[158 rows x 2 columns]

Le nombre moyen de votes pour le pays USA est environ 35016.7

3. Donner la liste des différents genres de films.

In [26]:
db.movies.distinct("genres")

['Action',
 'Adventure',
 'Animation',
 'Biography',
 'Comedy',
 'Crime',
 'Documentary',
 'Drama',
 'Family',
 'Fantasy',
 'Film-Noir',
 'History',
 'Horror',
 'Music',
 'Musical',
 'Mystery',
 'News',
 'Romance',
 'Sci-Fi',
 'Short',
 'Sport',
 'Talk-Show',
 'Thriller',
 'War',
 'Western']

Action est un genre de film.

4. Combien de films ont une note IMDb supérieure à 8 ?

In [28]:
db.movies.count_documents({
    "imdb.rating": { "$gt": 8 }
})

1177

Il y a 1177 films qui ont une note IMDb supérieure à 8.

5. Afficher le titre, l'année de sortie et les réalisateurs de tous les films (ne pas afficher l'ID).

In [30]:
Q5 = db.movies.find(
    {},
    { "_id": 0, "title": 1, "year" : 1, "directors" : 1}
)
pd.DataFrame(list(Q5))

title  \
0      Intolerance: Love's Struggle Throughout the Ages   
1                             The Poor Little Rich Girl   
2                                               The Kid   
3                                        Miss Lulu Bett   
4                                          Safety Last!   
...                                                 ...   
23525                                            Emelie   
23526                                    Techo y comida   
23527                                             Nahid   
23528                                  Coin Locker Girl   
23529                                 Tryapichnyy soyuz   

                            directors  year  
0                     [D.W. Griffith]  1916  
1                  [Maurice Tourneur]  1917  
2                   [Charles Chaplin]  1921  
3               [William C. de Mille]  1921  
4      [Fred C. Newmeyer, Sam Taylor]  1923  
...                               ...   ...  
23525                [Michael Thelin]  2015  
23526      [Juan Miguel del Castillo]  2015  
23527                [Ida Panahandeh]  2015  
23528                   [Jun-hee Han]  2015  
23529            [Mikhail Mestetskiy]  2015  

[23530 rows x 3 columns]

Charles Chaplin a réalisé The Kid en 1921.

6. Afficher le titre et le plot des films réalisés par D.W. Griffith.

In [90]:
Q6 = db.movies.find(
    { "directors": "D.W. Griffith"},
    { "_id": 0, "title": 1, "plot" : 1, "directors" : 1}
)
pd.DataFrame(list(Q6))
# A revoir !

plot  \
0  The story of a poor young woman, separated by ...   
1  A greedy tycoon decides, on a whim, to corner ...   
2  A frail waif, abused by her brutal boxer fathe...   
3  The Civil War divides friends and destroys fam...   

                                              title        directors  
0  Intolerance: Love's Struggle Throughout the Ages  [D.W. Griffith]  
1                                 A Corner in Wheat  [D.W. Griffith]  
2    Broken Blossoms or The Yellow Man and the Girl  [D.W. Griffith]  
3                             The Birth of a Nation  [D.W. Griffith]

D.D.W. Griffith a réalisé A Corner in Wheat	dont le plot est "A greedy tycoon decides, on a whim, to corner ...	".

7. Quels sont les 5 acteurs les plus populaires (ceux qui ont joué dans le plus de films) ?

In [61]:
Q7 = db.movies.aggregate([
    { "$unwind": "$cast" },
    { "$group": { "_id": "$cast", "nb": { "$sum": 1 }}},
    { "$sort": { "nb": -1 }},
    { "$limit": 5}
])
pd.DataFrame(list(Q7))

_id  nb
0      Gèrard Depardieu  68
1        Robert De Niro  60
2         Michael Caine  52
3  Marcello Mastroianni  50
4          Bruce Willis  49

Gerard Depardieu a joué dans 68 films, ce qui fait de lui l'acteur le plus populaire.

8. Quelles sont les 6 années avec le plus de films produits ?

In [64]:
Q8 = db.movies.aggregate([
    { "$group": { "_id": "$year", "nb": { "$sum": 1 }}},
    { "$sort": { "nb": -1 }},
    { "$limit": 6}
])
pd.DataFrame(list(Q8))

_id    nb
0  2013  1220
1  2014  1147
2  2012  1109
3  2011  1040
4  2009   990
5  2010   970

Il y a eu 1220 films produit en  2013, c'est l'année la plus productive.

9. Combien de films ont été réalisés avant l'année 2000 et ont une note IMDb supérieure à 7 ?

In [71]:
db.movies.count_documents({
    "year": {"$lt": 2000 },
    "imdb.rating": { "$gt": 7 }
})

4747

Il y a 4747 films réalisés avant l'année 2000 et ont une note IMDb supérieure à 7.

10. Donner les 20 réalisateurs qui ont réalisé le plus de films (afficher seulement leur nom et le nombre de films).

In [92]:
Q10 = db.movies.aggregate([
    { "$unwind": "$directors" },
    { "$group": { "_id": "$directors", "nb": { "$sum": 1 }}},
    { "$sort": { "nb": -1 }},
    { "$limit": 20}
])
pd.DataFrame(list(Q10))

_id  nb
0         Woody Allen  40
1           John Ford  35
2         John Huston  34
3       Takashi Miike  34
4       Werner Herzog  33
5     Martin Scorsese  32
6    Alfred Hitchcock  31
7        Sidney Lumet  30
8    Steven Spielberg  29
9     Mario Monicelli  29
10      Michael Apted  29
11       George Cukor  29
12      Robert Altman  28
13  Steven Soderbergh  28
14          Spike Lee  28
15    Jean-Luc Godard  27
16     Clint Eastwood  27
17          Ken Loach  27
18         Johnnie To  27
19        Wim Wenders  27

Woody Allen a réalisé 40 films, c'est celui qui en a fait le plus.

11. Quelle est la note IMDb moyenne des films sortis après 2010 ?

In [149]:
Q11 = db.movies.aggregate([
    {"$match": {"year": {"$gt": 2010 }}},
    { "$group": { "_id": "", "note_moyenne": { "$avg": "$imdb.rating" }}},
    { "$project": {"_id": 0, "note_moyenne": 1}}
])
pd.DataFrame(list(Q11))

note_moyenne
0      6.544033

La note IMDb moyenne des films sortis après 2010 est environ 6.5

12. (bonus) Afficher les films dont le titre contient le mot "love".

In [321]:
Q12 = db.movies.find( 
    { "title": { "$regex": "love(?!\S)", "$options": "i" } },
    { "_id": 0, "title": 1}
)
pd.DataFrame(list(Q12))

title
0                     Arise, My Love
1                  One Night of Love
2              Love Finds Andy Hardy
3                    Love Me Tonight
4                   Love on the Dole
..                               ...
316                             Love
317  Bang Gang (A Modern Love Story)
318                   Valley of Love
319              Cafè. Waiting. Love
320   Larry Kramer in Love and Anger

[321 rows x 1 columns]

On a 321 films qui contient le mot "love".

13. (bonus) Afficher le titre des films ayant un score Rotten Tomatoes (viewer) supérieur à 75.


In [309]:
Q13 = db.movies.find(
    {"tomatoes.rotten": { "$gt": 75 }},
    { "_id": 0, "title": 1}
)
pd.DataFrame(list(Q13))

title
0                   Glitter
1            Batman & Robin
2              The Avengers
3       Eye of the Beholder
4            Fantastic Four
..                      ...
712                 Taken 3
713                Blackhat
714  Magic in the Moonlight
715               Insurgent
716             Hot Pursuit

[717 rows x 1 columns]

Le film Glitter a un score Rotten Tomatoes (viewer) supérieur à 75.